glommio/src/sys/source.rs

```diff
@@ -225,3 +239,112 @@ impl Drop for Source {
+}
+
+impl fmt::Debug for SharedSource {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("InnerSharedSource")
+            .field("source", &self.source)
+            .field("wakers", &self.wakers)
+            .field("result", &self.result)
+            .finish()
+    }
+}
+
+impl From<Source> for SharedSource {
+    fn from(src: Source) -> Self {
+        SharedSource {
+            source: src,
+            result: RefCell::new(None),
+            wakers: RefCell::new(Default::default()),
+        }
+    }
+}
+
+impl SharedSource {
+    pub(crate) fn source_type(&self) -> Ref<'_, SourceType> {
+        self.source.source_type()
+    }
+
+    pub(crate) fn buffer(&self) -> Ref<'_, IoBuffer> {
+        Ref::map(self.source_type(), |stype| match &*stype {
+            SourceType::Read(_, Some(buffer)) => buffer,
+            SourceType::Write(_, buffer) => buffer,
+            x => panic!("Could not extract buffer. Source: {:?}", x),
+        })
+    }
+
+    pub(crate) unsafe fn as_slice(&self) -> &[u8] {
+        match &*self.source_type() {
+            SourceType::Read(_, Some(buffer)) => {
+                std::slice::from_raw_parts(buffer.as_ptr(), buffer.len())
+            }
+            SourceType::Write(_, buffer) => {
+                std::slice::from_raw_parts(buffer.as_ptr(), buffer.len())
+            }
+            x => panic!("Could not extract buffer. Source: {:?}", x),
+        }
+    }
+
+    pub(crate) fn add_waiter(&self, waker: Waker) {
+        // add the waker to the underlying source and insert the previous one (if any)
+        // into the backlog.
+        if let Some(previous) = self.source.add_waiter(waker) {
+            // before inserting the waker make we replace an old one if any
+            // TODO(Hippo): this is a linear scan so it could become problematic if a large
+            // number of tasks are all subscribing to this. For now, good enough.
+            let mut wakers = self.wakers.borrow_mut();
+            if let Some(old) = wakers.iter_mut().find(|x| x.will_wake(&previous)) {
+                *old = previous;
+            } else {
+                wakers.push_back(previous);
+            }
+        }
+    }
+
+    fn wake_waiters(&self) {
+        let wakers = self.wakers.replace(Default::default());
+        wakers.into_iter().for_each(Waker::wake);
+    }
+
+    pub(crate) fn result(&self) -> Option<io::Result<usize>> {
+        let res = if self.result.borrow().is_some() {
+            // we have extracted the result already, return that
+            self.result.borrow().clone().map(|x| x.into())
+        } else if let Some(result) = self.source.take_result() {
+            // the underlying source has a result, extract it
+            self.result.replace(Some(result.into()));
+            self.result.borrow().map(|x| x.clone().into())
+        } else {
+            // no result yet
+            None
+        };
+
+        if res.is_some() {
```

Comment:
it's strange that this happens at the time we capture the result.

What is your rationale here ?

Reply:
When a `Source` completes the reactor will wake the waker registered for it. For `SharedSource` I need this task to cooperate and wake the others. This is what this is doing.

I could explicitly call `wake_waiters` in consuming code but I felt it could remain an implementation detail hidden to consumers. This way, they can't get it wrong since this is the only way they can get the result out of the `SharedSource`.